src/Component/State/BulkAwareProcessor.php

```diff
@@ -0,0 +1,48 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Resource\State;
+
+use Sylius\Component\Resource\Context\Context;
+use Sylius\Component\Resource\Metadata\BulkOperationInterface;
+use Sylius\Component\Resource\Metadata\Operation;
+
+/**
+ * @experimental
+ */
+final class BulkAwareProcessor implements ProcessorInterface
+{
+    public function __construct(
+        private ProcessorInterface $decorated,
+    ) {
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function process(mixed $data, Operation $operation, Context $context): mixed
+    {
+        if (
+            !$operation instanceof BulkOperationInterface ||
+            !\is_iterable($data)
+        ) {
+            return $this->decorated->process($data, $operation, $context);
+        }
+
+        foreach ($data as $item) {
+            $this->decorated->process($item, $operation, $context);
```

Comment:
Can you confirm, that for bulk operations we are still triggering pre and post events for each item in bulk? Also, when do we flush in such situation

Reply:
Yes, we trigger a pre and post events for each item. There is a EventDispatcherBulkAwareProcessor with do that.
I've checked and everything seems ok, but we can check together before releasing the alpha. Maybe we can add some logic into the "Smoke subscriber" to log that and check if everything is ok.